sqlanalytics/resource_sql_endpoint.go

```diff
@@ -0,0 +1,185 @@
+)
+
+// ClusterSizes for SQL endpoints
+var (
+	ClusterSizes   = []string{"2X-Small", "X-Small", "Small", "Medium", "Large", "X-Large", "2X-Large", "3X-Large", "4X-Large"}
+	MaxNumClusters = 30
+)
+
+// SQLEndpoint ...
+type SQLEndpoint struct {
+	ID                 string      `json:"id,omitempty" tf:"computed"`
+	Name               string      `json:"name"`
+	ClusterSize        string      `json:"cluster_size"`
+	AutoStopMinutes    int         `json:"auto_stop_mins,omitempty"`
+	MinNumClusters     int         `json:"min_num_clusters,omitempty"`
+	MaxNumClusters     int         `json:"max_num_clusters,omitempty"`
+	NumClusters        int         `json:"num_clusters,omitempty"`
+	EnablePhoton       bool        `json:"enable_photon,omitempty"`
+	InstanceProfileARN string      `json:"instance_profile_arn,omitempty"`
+	State              string      `json:"state,omitempty" tf:"computed"`
+	JdbcURL            string      `json:"jdbc_url,omitempty" tf:"computed"`
+	OdbcParams         *OdbcParams `json:"odbc_params,omitempty" tf:"computed"`
+	Tags               *Tags       `json:"tags,omitempty"`
+}
+
+// OdbcParams ...
+type OdbcParams struct {
+	Host     string `json:"host"`
+	Path     string `json:"path"`
+	Protocol string `json:"protocol"`
+	Port     int32  `json:"port"`
+}
+
+// Tags ...
+type Tags struct {
+	CustomTags []Tag `json:"custom_tags"`
+}
+
+// Tag ...
+type Tag struct {
+	Key   string `json:"key"`
+	Value string `json:"value"`
+}
+
+// EndpointList ...
+type EndpointList struct {
+	Endpoints []SQLEndpoint `json:"endpoints"`
+}
+
+// NewSQLEndpointsAPI ...
+func NewSQLEndpointsAPI(ctx context.Context, m interface{}) SQLEndpointsAPI {
+	return SQLEndpointsAPI{m.(*common.DatabricksClient), ctx}
+}
+
+// SQLEndpointsAPI ...
+type SQLEndpointsAPI struct {
+	client  *common.DatabricksClient
+	context context.Context
+}
+
+// List all SQL endpoints
+func (a SQLEndpointsAPI) List() (l EndpointList, err error) {
+	err = a.client.Get(a.context, "/sql/endpoints", nil, &l)
+	return
+}
+
+// Start ..
+func (a SQLEndpointsAPI) Start(endpointID string, timeout time.Duration) error {
+	err := a.client.Post(a.context, fmt.Sprintf("/sql/endpoints/%s/start", endpointID), nil, nil)
+	if err != nil {
+		return err
+	}
+	return a.waitForRunning(endpointID, timeout)
+}
+
+// Stop ...
+func (a SQLEndpointsAPI) Stop(endpointID string) error {
+	return a.client.Post(a.context, fmt.Sprintf("/sql/endpoints/%s/stop", endpointID), nil, nil)
+}
+
+// Get ...
+func (a SQLEndpointsAPI) Get(endpointID string) (se SQLEndpoint, err error) {
+	err = a.client.Get(a.context, fmt.Sprintf("/sql/endpoints/%s", endpointID), nil, &se)
+	return
+}
+
+// Create ...
+func (a SQLEndpointsAPI) Create(se *SQLEndpoint, timeout time.Duration) error {
+	// maybe response should be something else...
+	err := a.client.Post(a.context, "/sql/endpoints", se, se)
+	if err != nil {
+		return err
+	}
+	return a.waitForRunning(se.ID, timeout)
```

Comment:
FYI: This may take a couple minutes b/c we wait for the Spark cluster to be running.